benchmarks/Lua/som.lua

```diff
@@ -0,0 +1,631 @@
+        local hash = key:custom_hash()
+        return bxor(hash, rshift(hash, 16))
+    end
+else
+    Dictionary.hash = assert(load[[
+    --  Lua 5.3 variant with bitwise operators
+    return function (self, key)
+        if not key then
+            return 0
+        end
+        local hash = key:custom_hash()
+        return hash ~ (hash >> 16)
+    end
+]])()
+end
+
+function Dictionary:is_empty ()
+    return self.size == 0
+end
+
+function Dictionary:get_bucket_idx (hash)
+    return (hash % self.buckets.n) + 1
+end
+
+function Dictionary:get_bucket (hash)
+    return self.buckets[self:get_bucket_idx(hash)]
+end
+
+function Dictionary:at (key)
+    local hash = self:hash(key)
+    local e = self:get_bucket(hash)
+    while e do
+        if e:match(hash, key) then
+            return e.value
+        end
+        e = e.next
+    end
+    return nil
+end
+
+function Dictionary:contains_key (key)
+    local hash = self:hash(key)
+    local e = self:get_bucket(hash)
+    while e do
+        if e.match(hash, key) then
+            return true
+        end
+        e = e.next
+    end
+    return false
+end
+
+function Dictionary:at_put (key, value)
+    local hash = self:hash(key)
+    local i = self:get_bucket_idx(hash)
+    local current = self.buckets[i]
+    if not current then
+        self.buckets[i] = self:new_entry(key, value, hash)
+        self.size = self.size + 1
+    else
+        self:insert_bucket_entry(key, value, hash, current)
+    end
+    if self.size > self.buckets.n then
+        self:resize()
+    end
+end
+
+function Dictionary:new_entry (key, value, hash)
+    return Entry.new(hash, key, value, nil)
+end
+
+function Dictionary:insert_bucket_entry (key, value, hash, head)
+    local current = head
+    while true do
+        if current:match(hash, key) then
+            current.value = value
+            return
+        end
+        if not current.next then
+            self.size = self.size + 1
+            current.next = self:new_entry(key, value, hash)
+            return
+        end
+        current = current.next
+    end
+end
+
+function Dictionary:resize ()
+    local old_storage = self.buckets
+    self.buckets = {n = old_storage.n * 2}
+    self:transfer_entries(old_storage)
+end
+
+function Dictionary:transfer_entries (old_storage)
+    local buckets = self.buckets
+    for i = 1, old_storage.n do
+        local current = old_storage[i]
+        if current then
+            old_storage[i] = nil
+            if not current.next then
+                local hash = (current.hash % buckets.n) + 1
+                buckets[hash] = current
+            else
+                self:split_bucket(old_storage, i, current)
+            end
+        end
+    end
+end
+
+function Dictionary:split_bucket (old_storage, i, head)
+    local lo_head, lo_tail = nil, nil
+    local hi_head, hi_tail = nil, nil
+    local current = head
+
+    while current do
+        if (floor(current.hash / old_storage.n) % 2) == 0 then
```

Comment:
`floor() % 2` is likely also a performance issue.
Perhaps a solution would be to use a helper method `bitAnd` that's specialized based on the Lua version.
That's only a small overhead in the interpreter, and for the compiler it should be easy to eliminate.